File: test/main/GraphTest.php
<?php
/**
 * Created by PhpStorm.
 * User: otakupasi
 * Date: 11/2/18
 * Time: 5:00 PM
 */

namespace main;

require "../../src/main/Graph.php";

use Navigator\Graph;
use PHPUnit\Framework\TestCase;

class GraphTest extends TestCase
{

    public function testSetOffset()
    {
        try {
            $graph = new Graph();
            $graph->addEdge(0, 1, 1);
            $this->assertEquals(0, $graph->getOffset(0));
            $graph->setOffset(0, 5);
            $this->assertEquals(5, $graph->getOffset(0));
        } catch (\Exception $exception) {
            $this->fail($exception->getTraceAsString());
        }
    }

    public function testHasEdge()
    {
        try {
            $graph = new Graph();
            $this->assertFalse($graph->hasEdge());
            $graph->addEdge(0, 1, 1);
            $this->assertTrue($graph->hasEdge());
        } catch (\Exception $exception) {
            $this->fail($exception->getTraceAsString());
        }
    }

    public function testGetWeight()
    {
        try {
            $graph = new Graph();
            $this->assertEquals(-1, $graph->getWeight(0, 1));
            $graph->addEdge(0, 1, 1);
            $this->assertEquals(1, $graph->getWeight(0, 1));
            $graph->addEdge(0, 2, 10);
            $this->assertEquals(10, $graph->getWeight(0, 2));
        } catch (\Exception $exception) {
            $this->fail($exception->getTraceAsString());
        }
    }

    public function testCountOutgoingEdges()
    {
        try {
            $graph = new Graph();
            $this->assertEquals(0, $graph->countOutgoingEdges(0));
            $graph->addEdge(0, 1, 1);
            $this->assertEquals(1, $graph->countOutgoingEdges(0));
            $graph->addEdge(1, 2, 1);
            $graph->addEdge(0, 2, 1);
            $this->assertEquals(2, $graph->countOutgoingEdges(0));
            $this->assertEquals(1, $graph->countOutgoingEdges(1));
        } catch (\Exception $exception) {
            $this->fail($exception->getTraceAsString());
        }
    }

    public function testGetOffset()
    {
        try {
            $graph = new Graph();
            $this->assertEquals(-1, $graph->getOffset(0));
            $graph->addEdge(0, 1, 1);
            $this->assertEquals(0, $graph->getOffset(0));
            $graph->addEdge(1, 2, 1);
            $this->assertEquals(1, $graph->getOffset(1));
            $graph->addEdge(0, 2, 2);
            $this->assertEquals(0, $graph->getOffset(0));
            $this->assertEquals(2, $graph->getOffset(1));
        } catch (\Exception $exception) {
            $this->fail($exception->getTraceAsString());
        }
    }

    public function testGetEdge()
    {
        try {
            $graph = new Graph();
            $this->assertFalse($graph->getEdge(0, 1));
            $graph->addEdge(0, 1, 1);
            $this->assertEquals(0, $graph->getEdge(0, 1));
            $graph->addEdge(1, 2, 1);
            $this->assertEquals(1, $graph->getEdge(1, 2));
            $graph->addEdge(0, 2, 2);
            $this->assertEquals(1, $graph->getEdge(0, 2));
            $this->assertEquals(2, $graph->getEdge(1, 2));
        } catch (\Exception $exception) {
            $this->fail($exception->getTraceAsString());
        }
    }

    public function testHasOutgoingEdges()
    {
        try {
            $graph = new Graph();
            $this->assertFalse($graph->hasOutgoingEdges(0));
            $graph->addEdge(0, 1, 1);
            $this->assertTrue($graph->hasOutgoingEdges(0));
            $this->assertFalse($graph->hasOutgoingEdges(1));
        } catch (\Exception $exception) {
            $this->fail($exception->getMessage());
        }
    }

    public function testNextNode()
    {
        $graph = new Graph();
        $this->assertEquals(-1, $graph->nextNode(0));
        $graph->addEdge(0, 1, 5);
        $this->assertEquals(1, $graph->nextNode(0));
        $graph->addEdge(0, 2, 10);
        $this->assertEquals(1, $graph->nextNode(0));

    }

    public function testAddEdge()
    {
        try {
            $graph = new Graph();
            $this->assertFalse($graph->hasEdge());
            $graph->addEdge(0, 1, 1);
            $this->assertTrue($graph->hasEdge());
            $this->assertTrue($graph->hasOutgoingEdges(0));
            $this->assertEquals(0, $graph->getEdge(0, 1));
            $graph->addEdge(1, 2, 1);
            $this->assertTrue($graph->hasOutgoingEdges(1));
            $this->assertEquals(1, $graph->getEdge(1, 2));
        } catch (\Exception $exception) {
            $this->fail($exception->getMessage());
        }
    }
}

File: test/main/GraphTest.java
package main;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

class GraphTest {

    @Test
    void runQuery() {
        //TODO
    }


//    @Test
//    void addEdge() {
//        try {
//            Graph graph = new Graph(3, 3);
//            assertFalse(graph.hasOutgoingEdges(0));
//            graph.addEdge(0, 1, 1);
//            assertTrue(graph.hasOutgoingEdges(0));
//            assertEquals(0, graph.getEdge(0, 1));
//            graph.addEdge(0, 2, 1);
//            graph.addEdge(1, 2, 1);
//            assertEquals(1, graph.getEdge(0, 2));
//            assertEquals(2, graph.getEdge(1, 2));
//        } catch (UnorderedGraphException e) {
//            fail("unordered graph exception");
//            e.printStackTrace();
//
//        }
//    }
//
//    @Test
//    void insertEdge() {
//        try {
//            Graph graph = new Graph(4, 3);
//            assertFalse(graph.hasOutgoingEdges(0));
//            graph.insertEdge(0, 1, 1);
//            assertTrue(graph.hasOutgoingEdges(0));
//            assertEquals(0, graph.getEdge(0, 1));
//            graph.insertEdge(1, 2, 1);
//            graph.insertEdge(0, 2, 1);
//            assertEquals(1, graph.getEdge(0, 2));
//            assertEquals(2, graph.getEdge(1, 2));
//        } catch (Exception e) {
//            fail("");
//        }
//    }

    @Test
    void hasOutgoingEdges() {
        try {
            Graph graph = new Graph(3, 3);
            assertFalse(graph.hasOutgoingEdges(0));
            graph.addEdge(0, 1, 1);
            assertTrue(graph.hasOutgoingEdges(0));
        } catch (Exception e) {
            fail("exception in hasOutgoing edges");
            e.printStackTrace();
        }
    }

    @Test
    public void getOffset() {
        try {
            Graph graph = new Graph(3, 3);
            assertEquals(-1, graph.getOffset(0));
            graph.addEdge(0, 1, 1);
            assertEquals(0, graph.getOffset(0));
        } catch (Exception e) {
            fail("exception in getOffset");
            e.printStackTrace();
        }
    }

    @Test
    public void setOffset() {
        try {
            Graph graph = new Graph(3, 3);
            assertEquals(-1, graph.getOffset(0));
            graph.setOffset(0, 0);
            assertEquals(0, graph.getOffset(0));
        } catch (Exception e) {
            fail("exception in setOffset");
            e.printStackTrace();
        }
    }

//    @Test
//    public void getEdge() {
//        try {
//            Graph graph = new Graph(3, 3);
//            assertEquals(-1, graph.getEdge(0, 1));
//            graph.addEdge(0, 1, 1);
//            assertEquals(0, graph.getEdge(0, 1));
//            graph.addEdge(0, 2, 1);
//            assertEquals(1, graph.getEdge(0, 2));
//        } catch (Exception e) {
//            fail("exception in getEdge");
//            e.printStackTrace();
//        }
//    }

    @Test
    public void nextNode() {
        try {
            Graph graph = new Graph(3, 3);
            assertEquals(-1, graph.nextNode(0));
            graph.addEdge(0, 1, 3);
            graph.addEdge(1, 2, 5);
            graph.addEdge(1, 0, 1);
            assertEquals(1, graph.nextNode(0));
            assertEquals(0, graph.nextNode(1));
        } catch (Exception e) {
            fail("exception in nextNode");
            e.printStackTrace();
        }
    }

    @Test
    public void countOutgoingEdges() {
        try {
            Graph graph = new Graph(3, 3);
            assertEquals(0, graph.countOutgoingEdges(0));
            graph.addEdge(0, 1, 1);
            graph.addEdge(0, 2, 2);
            assertEquals(2, graph.countOutgoingEdges(0));
        } catch (Exception e) {
            fail("exception");
            e.printStackTrace();
        }
    }

//    @Test
//    public void getWeight() {
//        try {
//            Graph graph = new Graph(3, 3);
//            assertEquals(-1, graph.getWeight(0, 1), 0.1);
//            graph.addEdge(0, 1, 4);
//            assertEquals(4, graph.getWeight(0, 1), 0.1);
//        } catch (Exception e) {
//            fail("exception");
//            e.printStackTrace();
//        }
//    }

    @Test
    public void getDistance() {
        //also tests functionality of setDistance
        try {
            Graph graph = new Graph(3, 3);
            assertEquals(Double.POSITIVE_INFINITY, graph.getDistance(1), 0.1);
            graph.setDistance(1, 1.5);
            assertEquals(1.5, graph.getDistance(1), 0.1);
        } catch (Exception e) {
            fail("exception");
            e.printStackTrace();
        }
    }


    @Test
    public void getNodeList() {
        int[] initial = {-1, -1, -1};
        int[] arr = {0, 2, -1};
        try {
            Graph graph = new Graph(3, 3);
            assertArrayEquals(initial, graph.getNodeList());
            graph.addEdge(0, 1, 1);
            graph.addEdge(0, 2, 2);
            graph.addEdge(1, 2, 3);
            assertArrayEquals(arr, graph.getNodeList());
        } catch (Exception e) {
            fail("exception");
            e.printStackTrace();

        }
    }


//    @Test
//    void getWeight1() {
//        try {
//            Graph graph = new Graph(3, 3);
//            assertEquals(-1, graph.getWeight(graph.getEdge(0, 1)), 0.1);
//            graph.addEdge(0, 1, 4);
//            assertEquals(4, graph.getWeight(graph.getEdge(0, 1)), 0.1);
//        } catch (Exception e) {
//            fail("exception");
//            e.printStackTrace();
//        }
//
//    }
}